linera-base/src/crypto/mod.rs

```diff
@@ -213,6 +229,58 @@ impl TryFrom<&[u8]> for AccountSignature {
     }
 }
 
+/// A trait for signing keys.
+#[cfg_attr(not(web), async_trait)]
+#[cfg_attr(web, async_trait(?Send))]
+pub trait SigningKey {
+    /// The public key type corresponding to this signing key.
+    type PublicKey: Into<AccountPublicKey> + std::fmt::Debug;
+
+    /// Signs the given value and returns the signature.
+    fn sign<'a, 'b, A: BcsSignable<'b>>(&'a self, value: &A) -> AccountSignature;
```

Comment:
In general, I think that signing may update the internal state of the signer, so I'd be inclined to make this `&mut self`.

In the interest of keeping the concerns separate, I think it might be better to take the bytes here directly, so that this trait remains independent of `BcsSignable`.

Reply:
Also, since the wallet takes only one signer, the signer must support signing for multiple identities, since different chains may be signed by different keys, so this should take as parameter the identity we want to sign for.  (Alternatively the wallet could take a `HashMap<Identity, Box<dyn SigningKey>>` instead, but I think that's a less elegant solution.)

Reply:
> In the interest of keeping the concerns separate, I think it might be better to take the bytes here directly, so that this trait remains independent of `BcsSignable` (and, for example, we don't force redundant reserialization if the caller already has serialized bytes available).

Definitely a good point - although, as previously, this API is compatible with the already-existing `AccountSecretKey` type (which in turn was made compatible with the crypto we have and how we use `BcsSignable` types).

All in all, good point but orthogonal to this PR.

Reply:
> In general, I think that signing may update the internal state of the signer, so I'd be inclined to make this `&mut self`. We can always strengthen it later to `&self` if it turns out nothing does this, but it's much harder to go the other way.

I can't think of an example where that'd be the case. Hopefully that'll be never true as intuitively it could be dangerous - consective signatures shouldn't rely on previous ones as that could lead some secret (metadata).